### PR TITLE
[WIP] Events modelized with Metaballs using Canvas

### PR DESCRIPTION
**Pull request by benmoufm**
- [x] code to drag the text
- [x] supressing of the graph
- [x] supress useless comments
- [x] update tests
- [x] simplify code (in different little functions)

### CODE REVIEW

**Comment by fzaninotto** [comment]
Nice ! Please find a more meaningful name for your PR.

File: css/style.css [hunk 1]
```diff
@@ -26,3 +26,6 @@
     opacity: 0.1;
 }
 
+body {width : 900px ; position :relative ;}
```
Comment (fzaninotto) [inline css/style.css]: CS

File: lib/canvas.js [hunk 2]
```diff
@@ -0,0 +1,140 @@
+"use strict";
+/* global require, module, d3 */
+
+var configurable = require('./util/configurable');
+
+module.exports = function (d3) {
+
+	return function (config) {
```
Comment (fzaninotto) [inline lib/canvas.js]: CS, watch out for inentation

File: lib/canvas.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
+"use strict";
```
Comment (fzaninotto) [inline lib/canvas.js]: The file name is too generic. It's not a Canvas, it's a... what? It's part of the developer's job to find good names.

File: lib/canvas.js (hunk 2 above)
Comment (fzaninotto) [inline lib/canvas.js]: please reindent the entire file correctly

File: lib/canvas.js [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    config = config || {
+      xScale: null,
+      eventColor: null
+    };
+    for (var key in defaultConfig) {
+      config[key] = config[key] || defaultConfig[key];
+    }
+
+	var canvasHandler = function canvasHandler() {
+		var canvas = document.getElementsByTagName('canvas')[0];
```
Comment (fzaninotto) [inline lib/canvas.js]: nope, the canvas DOM element should be passed as parameter

File: lib/eventLine.js [hunk 5]
```diff
@@ -33,9 +33,26 @@ module.exports = function (d3) {
           .style('fill', 'black')
         ;
 
-        d3.select(this).selectAll('circle').remove();
+        //html
+        /*<canvas id ="mon_canvas" width="200" height="100" style="border:1px solid #000000;">
+    </canvas>*/
 
-        var circle = d3.select(this).selectAll('circle')
+        /*var canvas = d3.select(this)
```
Comment (fzaninotto) [inline lib/eventLine.js]: commented out code is verboten

File: lib/canvas.js [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+	canvas.width = 800; canvas.height = 600;
+	
+	window.onload = function(){
```
Comment (fzaninotto) [inline lib/canvas.js]: using `onload` isn't a good idea. Youre modifying a global object from a utility function.

File: lib/canvas.js [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    for (var key in defaultConfig) {
+      config[key] = config[key] || defaultConfig[key];
+    }
+
+	var canvasHandler = function canvasHandler() {
+		var canvas = document.getElementsByTagName('canvas')[0];
+	canvas.width = 800; canvas.height = 600;
+	
+	window.onload = function(){		
+		var ctx = canvas.getContext('2d');
```
Comment (fzaninotto) [inline lib/canvas.js]: call the variable `context`, it's even clearer.

File: lib/canvas.js [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+			ctx.font = "30px Arial";
+          ctx.textAlign = "center";
+          ctx.fillText("Toto",750/2,380/2);
+		}
+		redraw();
+		
+		var lastX=canvas.width/2, lastY=canvas.height/2;
+		var dragStart,dragged;
+		canvas.addEventListener('mousedown',function(evt){
```
Comment (fzaninotto) [inline lib/canvas.js]: every event listener you add, you must remove

File: lib/canvas.js [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+			document.body.style.mozUserSelect = document.body.style.webkitUserSelect = document.body.style.userSelect = 'none';
```
Comment (fzaninotto) [inline lib/canvas.js]: hackish

File: lib/canvas.js [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+				redraw();
+			}
+		},false);
+		canvas.addEventListener('mouseup',function(evt){
+			dragStart = null;
+			if (!dragged) zoom(evt.shiftKey ? -1 : 1 );
+		},false);
+
+		var scaleFactor = 1.1;
+		var zoom = function(clicks){
```
Comment (fzaninotto) [inline lib/canvas.js]: use `function zoom(clicks)` instead, it eases debugging since the function isn't anonymous anymore.

File: lib/canvas.js [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+		var lastX=canvas.width/2, lastY=canvas.height/2;
+		var dragStart,dragged;
+		canvas.addEventListener('mousedown',function(evt){
+			document.body.style.mozUserSelect = document.body.style.webkitUserSelect = document.body.style.userSelect = 'none';
+			lastX = evt.offsetX || (evt.pageX - canvas.offsetLeft);
+			lastY = evt.offsetY || (evt.pageY - canvas.offsetTop);
+			dragStart = ctx.transformedPoint(lastX,lastY);
+			dragged = false;
+		},false);
+		canvas.addEventListener('mousemove',function(evt){
```
Comment (fzaninotto) [inline lib/canvas.js]: Just like you did with the `zoom` function, it'd be a nice idea to extract the callback.

File: lib/canvas.js [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+
+		var handleScroll = function(evt){
+			var delta = evt.wheelDelta ? evt.wheelDelta/40 : evt.detail ? -evt.detail : 0;
+			if (delta) zoom(delta);
+			return evt.preventDefault() && false;
+		};
+		canvas.addEventListener('DOMMouseScroll',handleScroll,false);
+		canvas.addEventListener('mousewheel',handleScroll,false);
+	};
+	gkhead.src = 'http://phrogz.net/tmp/gkhead.jpg';
```
Comment (fzaninotto) [inline lib/canvas.js]: Not sure we nee that.

File: lib/canvas.js (hunk 2 above)
Comment (benmoufm) [inline lib/canvas.js]: this file is not needed for now, I wanted to try to separate the code for the canvas from the one used for the svg, but I couldn't make it work separately

File: lib/canvas.js [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+	ball.src   = 'http://phrogz.net/tmp/alphaball.png';
+	
+	// Adds ctx.getTransform() - returns an SVGMatrix
+	// Adds ctx.transformedPoint(x,y) - returns an SVGPoint
+	function trackTransforms(ctx){
```
Comment (fzaninotto) [inline lib/canvas.js]: This function is too long. Try to break it into smaller chunks.

File: lib/eventDrops.js [hunk 14]
```diff
@@ -61,6 +61,141 @@ module.exports = function (d3) {
         var graphHeight = data.length * 40;
         var height = graphHeight + config.margin.top + config.margin.bottom;
 
+
+        var canvas_width = 750;
```
Comment (fzaninotto) [inline lib/eventDrops.js]: should be configurable

File: lib/eventDrops.js [hunk 15]
```diff
@@ -61,6 +61,141 @@ module.exports = function (d3) {
...
+
+        var lastX=canvas_width/2, lastY=canvas_height/2;
+
+        d3.select(this).select('canvas').remove();
+        var canvas = d3.select(this)
+          .append('canvas')
+          .attr('id', "mon_canvas")
+          .attr('width', canvas_width)
+          .attr('height', canvas_height)
+          .attr('style', "border:1px solid #000000;");
```
Comment (fzaninotto) [inline lib/eventDrops.js]: I'd prefer that you add a class and use a stylesheet to style it

File: lib/eventDrops.js [hunk 16]
```diff
@@ -61,6 +61,141 @@ module.exports = function (d3) {
...
+
+        var canvas = document.getElementsByTagName('canvas')[0];
```
Comment (fzaninotto) [inline lib/eventDrops.js]: nope, this must be a parameter of the parent function (see `selection`)

File: lib/eventDrops.js [hunk 17]
```diff
@@ -61,6 +61,141 @@ module.exports = function (d3) {
...
+
+          ctx.font = "30px Arial";
+        ctx.textAlign = "center";
+        ctx.fillText("Toto",750/2,35);
+        ctx.fillText("Toto",750/2,75);
+        ctx.fillText("Toto",750/2,115);
+        ctx.fillText("Toto",750/2,155);
+
+        /*
+        // tracé des lignes du graph (a faire dans une boucle)
```
Comment (fzaninotto) [inline lib/eventDrops.js]: French comments are not in English (said Lapalisse)

File: lib/eventDrops.js [hunk 18]
```diff
@@ -61,6 +61,141 @@ module.exports = function (d3) {
...
+        ctx.beginPath();
```
Comment (fzaninotto) [inline lib/eventDrops.js]: commented out code. You have a Git, use it.

File: lib/eventDrops.js [hunk 19]
```diff
@@ -252,6 +393,64 @@ module.exports = function (d3) {
       });
     }
 
+    // Adds ctx.getTransform() - returns an SVGMatrix
+  // Adds ctx.transformedPoint(x,y) - returns an SVGPoint
+  function trackTransforms(ctx){
```
Comment (fzaninotto) [inline lib/eventDrops.js]: s/ctx/context

File: src/eventDrops.js [hunk 20]
```diff
@@ -1 +1,653 @@
...
\ No newline at end of file
+(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({"./lib/main.js":[function(require,module,exports){
```
Comment (fzaninotto) [inline src/eventDrops.js]: how come this is not minified?

**Comment by fzaninotto** [comment]
Please add a screenshot of your progress so far. An animated GIF is even better. Also, tests fail.

**Comment by benmoufm** [comment]
Tests fail because it tests the circles representing the data that were on the graph before

**Comment by fzaninotto** [comment]
That means you must update the tests, too!

File: css/style.css [hunk 21]
```diff
@@ -8,21 +8,30 @@
...
 .graph-body .line circle {
-    opacity: 0.1;
+	opacity: 0.1;
 }
 
+body {
+	width : 900px ; position :relative;
+} 
+img {
+	position : absolute ; top :0px ; left :0px;
```
Comment (jpetitcolas) [inline css/style.css]: Should not be in lib CSS. Should be relative to user choice, in its own CSS.

File: karma.conf.js [hunk 22]
```diff
@@ -0,0 +1,64 @@
+// Karma configuration
+// Generated on Fri Jun 19 2015 11:19:21 GMT+0200 (CEST)
+
+module.exports = function(config) {
```
Comment (jpetitcolas) [inline karma.conf.js]: To remove as you don't use Karma (as seen verbally with you).

File: lib/eventDrops.js [hunk 23]
```diff
@@ -3,257 +3,289 @@
...
-    eventZoom: null,
-    eventClick: null,
-    hasDelimiter: true,
-    hasTopAxis: true,
-    hasBottomAxis: function (data) {
-      return data.length >= 10;
-    },
-    eventLineColor: 'black',
-    eventColor: null
+		start: new Date(0),
```
Comment (ThieryMichel) [inline lib/eventDrops.js]: use space

**Comment by jpetitcolas** [comment]
Perfs are worst than before. Closing.

**Pull request closed without merging**